src/frontend/flight/sync/schema.rs

```diff
@@ -0,0 +1,152 @@
+        for (col_desc, field) in column_descriptors.iter().zip(schema.fields()) {
+            match ColumnRole::try_from(col_desc.role) {
+                Ok(ColumnRole::OldPk) => {
+                    old_pk_types.insert((&col_desc.name, field.data_type().clone()));
+                }
+                Ok(ColumnRole::NewPk) => {
+                    new_pk_types.insert((&col_desc.name, field.data_type().clone()));
+                }
+                Ok(ColumnRole::Value) => {}
+                Ok(ColumnRole::Changed) => {
+                    let err = if field.data_type() != &DataType::Boolean {
+                        format!(
+                            "Field for column with `Changed` role must be of type boolean: {}",
+                            field.name()
+                        )
+                    } else if field.is_nullable() {
+                        format!(
+                            "Field for column with `Changed` role can not be nullable: {}",
+                            field.name()
+                        )
+                    } else if !column_descriptors.iter().any(|other_cd| {
+                        col_desc.name == other_cd.name
+                            && other_cd.role == ColumnRole::Value as i32
+                    }) {
+                        format!(
+                            "Column with `Changed` role must point to an existing column with `Value` role: {}",
+                            &col_desc.name
+                        )
+                    } else {
+                        // All good
+                        continue;
+                    };
+                    return Err(SyncError::SchemaError { reason: err });
+                }
+                Err(err) => {
+                    return Err(SyncError::SchemaError {
+                        reason: format!("Failed parsing role: {err:?}"),
+                    });
+                }
+            }
+        }
+
+        if old_pk_types.is_empty() || new_pk_types.is_empty() {
+            return Err(SyncError::SchemaError {
+                reason: "Change requested but batches do not contain old/new PK columns"
+                    .to_string(),
+            });
+        }
+
+        if old_pk_types != new_pk_types {
+            return Err(SyncError::SchemaError {
+                reason: "Change requested but old and new PK columns are not the same"
+                    .to_string(),
+            });
+        }
+
+        let columns = column_descriptors
+            .iter()
+            .zip(schema.fields())
+            .map(|(column_descriptor, field)| SyncColumn {
+                role: column_descriptor.role(),
+                name: column_descriptor.name.clone(),
+                field: field.clone(),
+            })
+            .collect();
+
+        Ok(Self { schema, columns })
+    }
+
+    #[allow(dead_code)]
+    pub fn arrow_schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
```

Comment:
Why keep this code if it's dead, some future impl?

Reply:
Potentially, but also I use it in the tests.

Reply:
You can make it non-public then (it'll still be available to unit tests in the same file)

Reply:
The unit test is in another file, so it's simpler to remove and revise the test setup slightly.